NameResolutionPal.Unix enabled async name resolution

gfoidl opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/33378

Also

* added basic tests in `System.Net.NameResolution.Pal.Tests`
* same order of methods as in the Windows-Pal

---
Edit: this issue is fixed.

Marking this PR as draft, as there is some bug anywhere which I can't find.
To check the native implementation 8c955bc4bd1d0dbe4b2823f4327ad52bf9856fd7 adds some (c)tests -- this will be reverted later -- and they pass without any problem.
The managed implementation is analogous to the windows' one.

The bug causes either a hang or a segfault.
For the latter I can see GDB Output (see [core-dump.txt](https://github.com/dotnet/runtime/files/4443296/core-dump.txt)).

Maybe it's a dumb failure on my side, which I don't see atm, or something different.

If the test (added with 4614267d909794899188d28ae05bddcdd53e7b08) doesn't hang, the native side it will call into the OS with return code `0`. The crash happens anywhere after the OS-call and the invocation of the callback. Unfortunately I'm not able to debug it further (due to lack of knowledge here deep down that stack).

gfoidl committed:
Native implementation

gfoidl committed:
Native tests

gfoidl committed:
Managed implementation

gfoidl committed:
Managed tests

gfoidl commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetHostName.cs:
```diff
@@ -2,39 +2,13 @@
         [DllImport(Libraries.SystemNative, EntryPoint = "SystemNative_GetHostName", SetLastError = true)]
-        private static extern unsafe int GetHostName(byte* name, int nameLength);
-
-        internal static unsafe string GetHostName()
```
Moved this to `NameResolutionPal.Unix`, as it is in the windows pal.

gfoidl commented on src/libraries/Native/Unix/CMakeLists.txt:
```diff
@@ -210,3 +210,5 @@ endif()
     add_subdirectory(System.Security.Cryptography.Native.Apple)
 endif()
+
+enable_testing()
```
These native tests will be removed later.

gfoidl reviewed (commented):
Minor notes for review

gfoidl committed:
Revert Interop.GetHostName change -- it's needed at other places too

So this change was a bad idea ;-)

gfoidl committed:
Fixed builds failures due to "unused parameter"

scalablecory commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -432,6 +422,137 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+        return GetAddrInfoErrorFlags_EAI_BADARG;
+    }
+
+    struct addrinfo hint;
+    GetHostEntryForNameCreateHint(&hint);
```
This `hint` structure will not last the length of the async operation. I'm not sure if it is required to.

gfoidl replied:
Hm, then the native tests (added temporarily with this PR) should fail too?!

Anyway, I'll change this to be on the safe side and the struct lasts the whole operation (will pack it into `GetAddrInfoAsyncState`).

Maybe that's the obvious failure I didn't see. Thanks therefore.

gfoidl replied:
Unfortunately no change when `hint` isn't on the stack but a member of the heap-allocation struct `GetAddrInfoAsyncState`. Still crashes.

lpereira replied:
This is const, so could be a global static constant. (See other comment.)

gfoidl committed:
Native: move struct addrinfo hint from stack-alloc to heap-alloc

Before I didn't last the async operation as it was on the stack. Now it's part of the GetAddrInfoAsyncState which gets heap allocted, so it last.
I'm not sure if addrinfo needs to last the async operation, as the native tests pass either way. This change makes it more correct, nevertheless.

lpereira commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -432,6 +422,137 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+}
+#endif
+
+static void GetHostEntryForNameCreateHint(struct addrinfo* hint)
```
Since the parameter to `getaddrinfo()` is constant, this could be defined statically and the address passed like that everywhere where it's needed:

```suggestion
static const struct addrinfo s_hostEntryForNameCreateHint = { .ai_family = AI_UNSPEC, .ai_flags = AI_CANONNAME };
```

lpereira commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -432,6 +422,137 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+}
+#endif
+
+static void GetHostEntryForNameCreateHint(struct addrinfo* hint)
+{
+    // Get all address families and the canonical name
+
+    memset(hint, 0, sizeof(struct addrinfo));
+    hint->ai_family = AF_UNSPEC;
+    hint->ai_flags = AI_CANONNAME;
+}
```
Since the parameter to `getaddrinfo()` is constant, this could be defined statically and the address passed like that everywhere where it's needed:

```suggestion
static const struct addrinfo s_hostEntryForNameCreateHint = { .ai_family = AI_UNSPEC, .ai_flags = AI_CANONNAME };
```

lpereira commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -432,6 +422,137 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+
+    GetHostEntryForNameCallback callback = state->callback;
+
+    int result = gai_error(&state->gai_request);
+    HostEntry* entry = state->entry;
+
+    int ret = ConvertGetAddrInfoAndGetNameInfoErrorsToPal(result);
+
+    if (result == 0)
+    {
+        const uint8_t* address = state->address;
+        struct addrinfo* info = state->gai_request.ar_result;
+
+        ret = GetHostEntries(address, info, entry);
+    }
+
+    free(state);
+
+    if (callback != NULL)
+    {
+        callback(entry, ret);
+    }
```
```suggestion
    int ret = ConvertGetAddrInfoAndGetNameInfoErrorsToPal(gai_error(&state->gai_request));

    if (ret == 0)
    {
        const uint8_t* address = state->address;
        struct addrinfo* info = state->gai_request.ar_result;

        ret = GetHostEntries(address, info, entry);
    }

    if (callback != NULL)
    {
        callback(state->entry, ret);
    }

    free(state);
```

gfoidl replied:
I had the `free(state)` intentionally before calling the callback.

* `entry` is passed from managed to native and via callback back to managed -- so the position of `free(state)` doesn't change anything
* if the callback (managed side) has an exception `state` won't be freed an so this can be a memory leak -- although `ProcessResult` on the managed side (the "callback worker") has a `try-finally`

I prefer leaving the `free(state)` before the callback. When there is a good reason to move it after the callback, I'll do it for sure.

lpereira replied:
Both are equivalent, yes, but the suggested change is slightly shorter (and doesn't have a variable that's attributed to but never read from, which is cleaner).

gfoidl replied:
OK.

lpereira replied:
(It's just a suggestion, no need to change it at all.)

gfoidl replied:
Yeah, but I like it as the code looks cleaner.
Anyway I highly appreciate the feedback here! (My C is a bit rusty, so it's good to get some hints.)

lpereira commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -432,6 +422,137 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+    hint->ai_flags = AI_CANONNAME;
+}
+
+int32_t SystemNative_PlatformSupportsGetAddrInfoAsync()
+{
+#if HAVE_GETADDRINFO_A
+    return 1;
+#else
+    return 0;
+#endif
+}
```
Is there a way to avoid this P/Invoke every time?

Also, since cmakedefine01 is used, maybe this could be rewritten as:

```suggestion
int32_t SystemNative_PlatformSupportsGetAddrInfoAsync()
{
    return HAVE_GETADDRINFO_A;
}
```

gfoidl replied:
It's cached on the managed side.

Nice trick to return the define. I didn't think about that. Thx. 

Will address the other points tomorrow.

lpereira commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -432,6 +422,137 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+
+int32_t SystemNative_GetHostEntryForNameAsync(const uint8_t* address, HostEntry* entry, GetHostEntryForNameCallback callback)
+{
+#if HAVE_GETADDRINFO_A
```
Shouldn't this return an error code if HAVE_GETADDRINFO_A isn't set to 1?

lpereira commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -432,6 +422,137 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+    struct addrinfo hint;
+    GetHostEntryForNameCreateHint(&hint);
+
+    struct GetAddrInfoAsyncState* state = malloc(sizeof(struct GetAddrInfoAsyncState));
+
+    state->gai_request.ar_name = (const char*)address;
+    state->gai_request.ar_service = NULL;
+    state->gai_request.ar_request = &hint;
+    state->gai_request.ar_result = NULL;
+    state->gai_requests = &state->gai_request;
+
+    state->sigevent.sigev_notify = SIGEV_THREAD;
+    state->sigevent.sigev_value.sival_ptr = state;
+    state->sigevent.sigev_notify_function = GetHostEntryForNameAsyncComplete;
+
+    state->address = address;
+    state->entry = entry;
+    state->callback = callback;
```
```suggestion
    struct GetAddrInfoAsyncState* state = malloc(sizeof(*state));

    *state = (struct GetAddrInfoAsyncState) {
        .gai_request = {
             .ar_name = ...,
             .ar_service = ...,
             ...,
        },
        .sigevent = {
            .sigev_notify = ...,
            ...,
         },
         .address = ...,
         ...
    };
```

lpereira commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -432,6 +422,137 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+
+    return result;
+#else
+    // Actually not needed, but otherwise the parameters are unused.
+    if (address != NULL && entry != NULL && callback != NULL)
+    {
+        return -1;
+    }
```
```suggestion
    (void)address;
    (void)entry;
    (void)calback;
```

lpereira commented on src/libraries/Native/Unix/System.Native/tests.c:
```diff
@@ -0,0 +1,86 @@
+#include "pal_networking.h"
```
License header missing?

gfoidl replied:
This file / commit will be reverted and won't be part of the final PR.
It's just for narrowing down the bug. These tests pass...

lpereira commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -432,6 +422,137 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+        return ConvertGetAddrInfoAndGetNameInfoErrorsToPal(result);
+    }
+
+    return result;
```
Shouldn't result be converted to a PAL code too?

gfoidl replied:
It's 0 anyway.

lpereira commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -432,6 +422,137 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+{
+    struct GetAddrInfoAsyncState* state = (struct GetAddrInfoAsyncState*)context.sival_ptr;
+
+    atomic_thread_fence(memory_order_acquire);
```
Why is the fence necessary?

scalablecory replied:
This callback comes in on another thread. It ensures visibility to this thread of everything in `state`.

`getaddrinfo_a` probably does something like this already, so it might be safe to remove, but it isn't documented.

stephentoub commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:
```diff
@@ -2,24 +2,181 @@
+                    }
+                }
+            }
+        }
```
Why not just do this in a static field initializer?
```C#
private static readonly bool s_supportsGetAddrInfoAsync = Interop.Sys.PlatformSupportsGetAddrInfoAsync();
```
?

gfoidl replied:
Just to have it the same as the windows-implementation. But I see (now) that there is a need to start the socket, which isn't needed on linux.

I'll update to this.

gfoidl committed:
PR feedback

lpereira commented:
Are you still experiencing the crash/hang you mentioned in the PR message?

gfoidl commented:
Yes, unfortunately.

lpereira commented:
Have you tried either running under Valgrind or building with AddressSanitizer to have a better insight?

gfoidl commented:
With pure native execution the bug does not show up. Only when called via P/Invoke, i.e. from the managed unit test with testhost.

Do you know how to incorporate AddressSanitizer with the testhost?

lpereira commented:
I've never tested this with testhost, but it looks like if you build a debug build with the `$DEBUG_SANITIZERS` environment variable set to `asan`, it should create an AddressSanitizer-enabled build for you. (Looking in `src/coreclr/configurecompiler.cmake`, around line 117.)

gfoidl commented:
Thanks. Will try this tomorrow.

When I read the GDB dump (see attachment in top post) correct, I have the assumption that P/Invoke for Linux and threads has an issue. But I'm not really able to interpret this correctly, because of lack of knowledge with these internals. Hopefully that's not the case and P/Invoke is correct.

gfoidl commented:
Valgrind outputs quite some messages -- even for `dotnet new console` in a fresh docker container. I'd expect there shouldn't be only some message for OS-memory cleanup. Is this known / expected or should we file an issue to investigate and fix that?

Running `valgrind dotnet build /t:test` in `runtime/src/libraries/System.Net.NameResolution/tests/PalTests` the valgrind-output is a bit longer. 
So I removed the failling tests (the ones for the async code added with this PR), tests pass, and compared it to the valgrind-output for the failing tests (core dumps as described in the OP).
Except the process-id there's only one difference (leaving out the final heap summary):

```diff
-vex amd64->IR: unhandled instruction bytes: 0x48 0xE9 0xB0 0x35 0x42 0xC9 0xE8 0x2B
+vex amd64->IR: unhandled instruction bytes: 0x48 0xE9 0xB0 0x35 0x41 0xC9 0xE8 0x2B
 vex amd64->IR:   REX=1 REX.W=1 REX.R=0 REX.X=0 REX.B=0
 vex amd64->IR:   VEX=0 VEX.L=0 VEX.nVVVV=0x0 ESC=NONE
 vex amd64->IR:   PFX.66=0 PFX.F2=0 PFX.F3=0
-==xxxx== valgrind: Unrecognised instruction at address 0x3d9bafca.
-==xxxx==    at 0x3D9BAFCA: ???
-==xxxx==    by 0x3D41F58F: ??? (in /home/gfoidl/dotnet-repos/runtime/.dotnet/shared/Microsoft.NETCore.App/5.0.0-preview.3.20169.1/System.Private.CoreLib.dll)
-==xxxx==    by 0x3D443245: ??? (in /home/gfoidl/dotnet-repos/runtime/.dotnet/shared/Microsoft.NETCore.App/5.0.0-preview.3.20169.1/System.Private.CoreLib.dll)
-==xxxx==    by 0x3D45BCC2: ??? (in /home/gfoidl/dotnet-repos/runtime/.dotnet/shared/Microsoft.NETCore.App/5.0.0-preview.3.20169.1/System.Private.CoreLib.dll)
+==xxxx== valgrind: Unrecognised instruction at address 0x3d9cafca.
+==xxxx==    at 0x3D9CAFCA: ???
+==xxxx==    by 0x3D42F58F: ??? (in /home/gfoidl/dotnet-repos/runtime/.dotnet/shared/Microsoft.NETCore.App/5.0.0-preview.3.20169.1/System.Private.CoreLib.dll)
+==xxxx==    by 0x3D453245: ??? (in /home/gfoidl/dotnet-repos/runtime/.dotnet/shared/Microsoft.NETCore.App/5.0.0-preview.3.20169.1/System.Private.CoreLib.dll)
+==xxxx==    by 0x3D46BCC2: ??? (in /home/gfoidl/dotnet-repos/runtime/.dotnet/shared/Microsoft.NETCore.App/5.0.0-preview.3.20169.1/System.Private.CoreLib.dll)
 ==xxxx==    by 0x6DDD8BE: CallDescrWorkerInternal (in /home/gfoidl/dotnet-repos/runtime/.dotnet/shared/Microsoft.NETCore.App/5.0.0-preview.3.20169.1/libcoreclr.so)
 ==xxxx==    by 0x6D09CCA: MethodDescCallSite::CallTargetWorker(unsigned long const*, unsigned long*, int) (in /home/gfoidl/dotnet-repos/runtime/.dotnet/shared/Microsoft.NETCore.App/5.0.0-preview.3.20169.1/libcoreclr.so)
 ==xxxx==    by 0x6C5675D: CorHost2::_CreateAppDomain(char16_t const*, unsigned int, char16_t const*, char16_t const*, int, char16_t const**, char16_t const**, unsigned int*) (in /home/gfoidl/dotnet-repos/runtime/.dotnet/shared/Microsoft.NETCore.App/5.0.0-preview.3.20169.1/libcoreclr.so)
```

See attached files for raw output.

Will investigate further and try to incorparate some sanitizers.
[valgrind_new_console.txt](https://github.com/dotnet/runtime/files/4455835/valgrind_new_console.txt)
[valgrind_test_OK.txt](https://github.com/dotnet/runtime/files/4455844/valgrind_test_OK.txt)
[valgrind_test_KO.txt](https://github.com/dotnet/runtime/files/4455846/valgrind_test_KO.txt)

gfoidl committed:
Fixed leak in tests.c according to valgrind's run

gfoidl commented:
I have a clue now. The `address` is passed in as `string`, so the runtime needs to marshal it, and the marshalled string will only last until `SystemNative_GetHostEntryForNameAsync` returns. Then at the callback this reference has gone and it crashes.
This explains why the native tests pass, and the managed ones fail.

This can be seen in the following trace:
```
MANAGED: start with hostName: microsoft.com
MANAGED: addr context: 0x7f8dc8003560
MANAGED: addr state: 0x7f8e8f731220
NATIVE: arg address: microsoft.com
NATIVE: entry: 0x7f8dc8003560
NATIVE: callback: 0x7f8e13a93544
MANAGED: result from os-call: errorCode: 0
MANAGED: returning task
NATIVE (callback): entry: 0x7f8dc8003560
NATIVE (callback): callback: 0x7f8e13a93544
NATIVE (callback): address:  BÎwŽ
NATIVE (callback): gai_error: -2
NATIVE (callback): pal_error: 5
MANAGED: entering callback with error: 5
MANAGED: entering ProcessResult with errorCode: HostNotFound
MANAGED: addr context: 0x7f8dc8003560
MANAGED: addr state: 0x7f8e8f731220
```
In the callback the address is garbage.

A trivial workaround / fix would be duplicating the string in the native side (`strdup`).
Is there any .NET-way to tell P/Invoke how long the marshalled memory (string) should last?

stephentoub commented:
> Is there any .NET-way to tell P/Invoke how long the marshalled memory (string) should last?

No (and if there were, the best you could hope for would be "until it's manually free'd by the native code", since the runtime has no idea what the lifetime is if it's not tied to the synchronous method call's lifetime).  If you want to manage it yourself, you can just change the P/Invoke to take a `char*` and use something like `Marshal.StringToHGlobalAnsi`, then free it manually later.

gfoidl committed:
Fixed bug due to marshalled string argument

gfoidl commented:
The crash is gone, local tests are passing now.
I'm cleaning my branch up, add some tests, then "draft" will go away.

---

It's offtopic and a rare case, but ...

> and if there were, the best you could hope

... that there's an attribute that let's me specify the callback and so specify how long the marshaller should keep the memoy around.
Just a spontaneous idea, not thought to end.

gfoidl committed:
More managed PalTests

gfoidl committed:
Revert "Native tests"

This reverts commit 8c955bc4bd1d0dbe4b2823f4327ad52bf9856fd7 and d7cb2be7eb3ec09f27a537c89341fb63e176e0bc.

gfoidl commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:
```diff
@@ -2,24 +2,149 @@
+        public static unsafe string GetHostName() => Interop.Sys.GetHostName();
+
+        public static unsafe Task GetAddrInfoAsync(string hostName, bool justAddresses)
+        {
```
I see the case for `hostName == ""` is not handled. It's needed to be consistent to Windows.
Will add this plus a test therefore.

gfoidl commented on src/libraries/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs:
```diff
@@ -204,11 +213,145 @@ public void TryGetNameInfo_LocalHost_IPv6_TryGetAddrInfo(bool justAddresses)
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public async Task GetAddrInfoAsync_ExternalHost(bool justAddresses)
```
Should these tests be marked with `Outerloop`?

wfurt replied:
I probably would not at this point. Since all DNS tests depend on external server, we would need to make all of them outer loop IMHO. For now, I would assume that base resolution work and I would leave it as inner loop. was thinking while back about running batch of resolutions to make sure DNS work and perhaps prime caches so test runs are more predictable. I think we can leave is is for now and react if we see CI failures.

lpereira commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -432,12 +421,144 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+        return GetAddrInfoErrorFlags_EAI_MEMORY;
+    }
+
+    char* localAddress = strdup((const char*)address);
```
malloc() result is NULL-checked but strdup() isn't

lpereira replied:
You could do both allocations here in a single one by using a flexible member in struct GetAddrInfoAsyncState:

```c
struct GetAddrInfoAsyncState {
  ...
  char address[];
};

size_t addrlen = strlen(address);
struct GetAddrInfoAsyncState* state = malloc(sizeof(*state) + addrlen + 1);

memcpy(state->address, address, addlen+1);
```

gfoidl committed:
Handle the case of HostName="" and tests for this

gfoidl committed:
Use flexible array member to hold the address in struct state

Cf. https://github.com/dotnet/runtime/pull/34633#discussion_r406447963

gfoidl commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -432,12 +421,148 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+
+    size_t addrlen = strlen((const char*)address);
+
+    if (addrlen > _POSIX_HOST_NAME_MAX)
```
`_POSIX_HOST_NAME_MAX` is fixed to 255. It's more than the 253 according the RFC (cf. https://github.com/dotnet/runtime/pull/34633#discussion_r406447963), but it matches 
https://github.com/dotnet/runtime/blob/458579b71af2456bc97e410ecfb99219309f5cf5/src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetHostName.cs#L18

gfoidl replied:
BTW: neither the Windows-Pal nor the sync-version have a check for max hostname. Should these be added (at the managed side to catch them early)? Or leave it as the uncommon case and let the OS handle it?

lpereira replied:
In this case, it's better to check because you're computing the allocation length based on  what is potentially user input; that could overflow and become an issue. By capping it to `POSIX_HOST_NAME_MAX` like you did here, the overflow is avoided (and there's no need to perform an overflow check).

In other cases, it's fine to leave it to the OS to verify, as they're passing the string as it is to the system calls/library functions.

lpereira commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -432,12 +421,148 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+        .entry = entry,
+        .callback = callback
+    };
+
+    atomic_thread_fence(memory_order_release);
+
+    int32_t result = getaddrinfo_a(GAI_NOWAIT, &state->gai_requests, 1, &state->sigevent);
+
+    if (result != 0)
+    {
+        free(state);
+        return ConvertGetAddrInfoAndGetNameInfoErrorsToPal(result);
+    }
+
+    return result;
+#else
+    (void)address;
+    (void)entry;
+    (void)callback;
+
+    // GetHostEntryForNameAsync is not supported on this platform.
+    return -1;
+#endif
+}
+
 void SystemNative_FreeHostEntry(HostEntry* entry)
 {
     if (entry != NULL)
     {
-        free(entry->CanonicalName);
-        free(entry->IPAddressList);
+        if (entry->CanonicalName != NULL)
+        {
+            free(entry->CanonicalName);
+        }
+
+        if (entry->IPAddressList != NULL)
+        {
+            free(entry->IPAddressList);
+        }
```
Nit: You don't need to null-check before calling free().  free() can take any value returned by malloc(), including NULL.  (Change is harmless, but something to keep in mind.)

gfoidl replied:
Thanks for the valuable input!

Maybe it was an artefact of the bug hunt, but Valgrind complained this. That's why I changed this. Will revert to the cleaner version.

lpereira reviewed (approved):
Native side LGTM with a minor nit; can't speak about managed-side.

gfoidl committed:
Nits in native layer

Cf. https://github.com/dotnet/runtime/pull/34633#discussion_r406494734

gfoidl committed:
Merge branch 'master' into unix-async-name-resolution

stephentoub commented:
@karelz, this has been open for several months.  Can someone code review it?  Thanks.

karelz commented:
@gfoidl sorry for the delay. We should be able to get back to the code review in couple of weeks when we finish our most critical 5.0 bugs. Thanks for understanding and sorry for the XXXL delay :(

gfoidl commented:
@karelz no problem, I'm here :wink:

gfoidl committed:
Merge branch 'master' into unix-async-name-resolution

wfurt commented:
can you please sync up and resolve conflicts @gfoidl? Sorry for the delay but I think we are ready now so let's drive this to completion.

wfurt commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -364,6 +351,8 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
         }
     }
+#else
+    (void)address;
```
is this here to suppress warning about unused parameter in the else case?

gfoidl replied:
Yep. if `HAVE_GETIFADDRS` is not defined the C-compiler complains about unused parameters, hence this "trick" to avoid the warnings. Cf. https://github.com/dotnet/runtime/pull/34633#discussion_r405119626

wfurt commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -440,6 +429,135 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+#if HAVE_GETADDRINFO_A
+    if (address == NULL || entry == NULL)
+    {
+        return GetAddrInfoErrorFlags_EAI_BADARG;
```
It seems like this is later consumed by GetSocketErrorForNativeError. Should the be errno e.g. EINVAL instead of  GetAddrInfoErrorFlags_EAI_BADARG?

gfoidl replied:
I think `GetAddrInfoErrorFlags_EAI_BADARG` is correct, as it is mapped to https://github.com/dotnet/runtime/blob/e29b83934e7ceef9905f82bd6328cbc72db2278b/src/libraries/Common/src/Interop/Unix/System.Native/Interop.HostEntry.cs#L14-L24

wfurt commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -440,6 +429,135 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+        ret = GetHostEntries(address, info, state->entry);
+    }
+
+    if (callback != NULL)
```
is there valid case when the callback be null?

gfoidl replied:
No.

In C# I would write `Debug.Assert(callback is not null)`. Can we do something similar here or just drop the `if`?

gfoidl replied:
Will use
```c
    assert(callback != NULL);
```
instead.

wfurt reviewed (commented):
generally, looks good to me.

wfurt commented:
cc: @tmds

gfoidl commented:
> can you please sync up and resolve conflicts

Sure -- but next week...have a bit of work to do and unfortunately it's not just a simple merge-conflict now.
Will then address you comments too.

gfoidl committed:
Merge branch 'master' into unix-async-name-resolution

gfoidl commented:
https://github.com/dotnet/runtime/pull/33420 added support for "AddressFamily-specific name resolution".
This isn't covered in this PR so far. I think the async name resolution should provide this too. 
So add this to this PR or make this change separately?
Normally I'd prefer the separate change, but right now it's more like rewriting the PR on top of https://github.com/dotnet/runtime/pull/33420 so the AddressFamily-specific part could be done at the same time.

gfoidl committed:
Fixed native merge conflict + added AddressFamily-handling

gfoidl committed:
Fixed managed merge conflicts + added AddressFamily-handling

gfoidl committed:
Updated NameResolutionPalTests for AddressFamily

gfoidl committed:
Removed EnsureSocketsAreInitialized

On Unix this was a nop anyway, https://github.com/dotnet/runtime/pull/43284 removed it from Windows.

gfoidl committed:
Fixed bug at native side

Seems like "hint" must last the async operation, so stack-only won't do it.
Tests will crash with
```
The active test run was aborted. Reason: Test host process crashed : .../runtime/src/libraries/Native/Unix/System.Native/pal_networking.c (315): error -7: Unknown AddrInfo error flag. Unknown error -7 (0 failed)
```

If put into the heap, it succeeds.

gfoidl commented on src/libraries/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -519,6 +503,152 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, int32_t address
+    }
+
+    struct addrinfo hint;
+    memset(&hint, 0, sizeof(struct addrinfo));
```
I'm not sure if this zeroing is necessary. Tests will succeed either way, in the docs I can't find clear hints.
Being in doubt, it's safer to zero it.

Same in L606.

gfoidl replied:
Any I'll refactor this into one function, as this is done in the async code-path too.

gfoidl committed:
Refactor setting of the address family into TrySetAddressFamily

gfoidl committed:
Fixed unused parameter warning / failure

gfoidl commented:
@wfurt updated -- PTAL.

gfoidl commented:
The async version gets a CancellarionToken passed in. Exists on Unix a mechanism to cancel the async lookup? I'm not aware of, so ATM the token isn't used.

gfoidl commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:
```diff
@@ -1,23 +1,157 @@
+            return state.Task;
+        }
+
+        private static unsafe void GetHostEntryForNameCallback(Interop.Sys.HostEntry* entry, int error)
```
Should this callback (from native side) be attributed with UnmanagedCallersOnly?
```suggestion
        [UnmanagedCallersOnly]
        private static unsafe void GetHostEntryForNameCallback(Interop.Sys.HostEntry* entry, int error)
```

gfoidl commented:
The test `System.Net.NameResolution.Tests.LoggingTest.GetHostEntryAsync_InvalidHost_LogsError` fails.
The synchronous twin passes. With (poor mans) tracing can be seen:
```
====================================================================================================
SYNC
hostName: notahostname.invalid.corp.microsoft.com, justAddresses: False, addressFamily: Unspecified
result: 5
return value: HostNotFound
====================================================================================================
ASYNC
hostName: notahostname.invalid.corp.microsoft.com, justAddresses: False, addressFamily: Unspecified
in callback with error: 5
in processresult with errorCode: HostNotFound
set exception:
	HostNotFound
	System.Net.Sockets.SocketException (0xFFFDFFFF): Name or service not known
   at System.Net.NameResolutionPal.ProcessResult(SocketError errorCode, GetHostEntryForNameContext* context) in /home/gfoidl/dotnet-repos/runtime/src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:line 163
   at System.Net.NameResolutionPal.GetHostEntryForNameCallback(HostEntry* entry, Int32 error) in /home/gfoidl/dotnet-repos/runtime/src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:line 138
--- End of stack trace from previous location ---
```
So both fail with the same `HostNotFound`.

Need to investigate this further...

gfoidl commented:
With more tracing I see
```
====================================================================================================
SYNC
hostName: notahostname.invalid.corp.microsoft.com, justAddresses: False, addressFamily: Unspecified
result: 5
return value: HostNotFound
test, SocketErrorCode: HostNotFound
	e: System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (00000005, 0xFFFDFFFF): Name or service not known
   at System.Net.Dns.GetHostEntryOrAddressesCore(String hostName, Boolean justAddresses, AddressFamily addressFamily) in /home/gfoidl/dotnet-repos/runtime/src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:line 386
   at System.Net.Dns.GetHostEntryCore(String hostName, AddressFamily addressFamily) in /home/gfoidl/dotnet-repos/runtime/src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:line 367
   at System.Net.Dns.GetHostEntry(String hostNameOrAddress, AddressFamily family) in /home/gfoidl/dotnet-repos/runtime/src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:line 92
   at System.Net.Dns.GetHostEntry(String hostNameOrAddress) in /home/gfoidl/dotnet-repos/runtime/src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:line 61
   at System.Net.NameResolution.Tests.LoggingTest.<>c.<GetHostEntry_InvalidHost_LogsError>b__1_1() in /home/gfoidl/dotnet-repos/runtime/src/libraries/System.Net.NameResolution/tests/FunctionalTests/LoggingTest.cs:line 45
====================================================================================================
ASYNC
hostName: notahostname.invalid.corp.microsoft.com, justAddresses: False, addressFamily: Unspecified
in callback with error: 5
in processresult with errorCode: HostNotFound
set exception:
	HostNotFound
	System.Net.Sockets.SocketException (0xFFFDFFFF): Name or service not known
   at System.Net.NameResolutionPal.ProcessResult(SocketError errorCode, GetHostEntryForNameContext* context) in /home/gfoidl/dotnet-repos/runtime/src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:line 163
   at System.Net.NameResolutionPal.GetHostEntryForNameCallback(HostEntry* entry, Int32 error) in /home/gfoidl/dotnet-repos/runtime/src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:line 138
--- End of stack trace from previous location ---

test, SocketErrorCode: HostNotFound
	e: System.Net.Sockets.SocketException (0xFFFDFFFF): Name or service not known
   at System.Net.NameResolutionPal.ProcessResult(SocketError errorCode, GetHostEntryForNameContext* context) in /home/gfoidl/dotnet-repos/runtime/src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:line 163
   at System.Net.NameResolutionPal.GetHostEntryForNameCallback(HostEntry* entry, Int32 error) in /home/gfoidl/dotnet-repos/runtime/src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:line 138
--- End of stack trace from previous location ---
   at System.Net.NameResolution.Tests.LoggingTest.<>c.<GetHostEntryAsync_InvalidHost_LogsError>b__2_1() in /home/gfoidl/dotnet-repos/runtime/src/libraries/System.Net.NameResolution/tests/FunctionalTests/LoggingTest.cs:line 82
```
So in the sync version it's a `System.Net.Internals.SocketExceptionFactory+ExtendedSocketException` while the async version has a `System.Net.Sockets.SocketException`.

In the tests it's https://github.com/dotnet/runtime/blob/da39b2df168325d27a5af4c1fbf19474b4242fbf/src/libraries/System.Net.NameResolution/tests/FunctionalTests/LoggingTest.cs#L47 for sync, and https://github.com/dotnet/runtime/blob/da39b2df168325d27a5af4c1fbf19474b4242fbf/src/libraries/System.Net.NameResolution/tests/FunctionalTests/LoggingTest.cs#L82 for async.

https://github.com/dotnet/runtime/blob/da39b2df168325d27a5af4c1fbf19474b4242fbf/src/libraries/Common/src/System/Net/Internals/SocketExceptionFactory.cs#L11 so both (sync and async) hit this exception-filter (otherwise the trace couldn't show this info anyway), and the `SocketErrorCode` is the same (`HostNotFound`).

Now I just need to understand what going on here...why the sync test passes and async one doesn't. 
(Maybe it's too obvious now, but I don't see it).

gfoidl committed:
Little cleanup

gfoidl commented:
I believe this is an (existing) issue with the test code.

Before this PR it was async over sync on linux, and [windows test is disabled](https://github.com/dotnet/runtime/blob/da39b2df168325d27a5af4c1fbf19474b4242fbf/src/libraries/System.Net.NameResolution/tests/FunctionalTests/LoggingTest.cs#L68), but the same issue repros on windows if enabled.

Problem is that [OnEventWritten](https://github.com/dotnet/runtime/blob/5d2490edcdd5165795f509ca5dfb98e6516e1e3a/src/libraries/Common/tests/System/Diagnostics/Tracing/TestEventListener.cs#L124) never fires, so the test fails.

gfoidl commented:
(sorry for being noisy...)

Found (and fixed) the problem. 

https://github.com/dotnet/runtime/blob/5d2490edcdd5165795f509ca5dfb98e6516e1e3a/src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs#L128-L130
has no continuation for the faulted case, so nothing got logged.

gfoidl committed:
Fixed (unrelated) test failures

Cf. https://github.com/dotnet/runtime/pull/34633#issuecomment-746377633

gfoidl committed:
Made LoggingTest async instead of GetAwaiter().GetResult()

gfoidl commented on src/libraries/System.Net.NameResolution/tests/FunctionalTests/LoggingTest.cs:
```diff
@@ -66,29 +64,31 @@ public void GetHostEntry_InvalidHost_LogsError()
 
         [ConditionalFact]
         [PlatformSpecific(~TestPlatforms.Windows)]  // Unreliable on Windows.
```
Does this still hold? Locally I can't repro any problems with this test (now after 900834ed89229c5394e81d4316fdc20fd3daa78d)

This got added in https://github.com/dotnet/corefx/pull/42428

Just want to double-check is this test revealed a miss in product-code.

wfurt replied:
I'm not sure. Could be use CI stability while back. Perhaps best would be do this separately e.g. put this in and do separate PR to enable this in ci and watch the outcome. If we see failures, we can try to reproduce locally or we can try to collect more info.

gfoidl commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.HostEntry.cs:
```diff
@@ -32,8 +32,16 @@ internal unsafe struct HostEntry
             internal int IPAddressCount;       // Number of IP addresses in the list
         }
 
+        internal unsafe delegate void GetHostEntryForNameCallback(HostEntry* entry, int status);
```
Should this be a function instead?

(In the meantime we got C# 9 😃).

gfoidl committed:
Use function pointer

gfoidl commented:
From [CI](https://dev.azure.com/dnceng/public/_build/results?buildId=923863&view=ms.vss-test-web.build-test-results-tab&runId=29322842&resultId=116061&paneView=debug): this test https://github.com/dotnet/runtime/blob/f85c316051f7ba53e54349620e09e034dbd1a812/src/libraries/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs#L51 fails on `net6.0-Android-Release-arm64-Mono_Release-Windows.10.Amd64.Android.Open`

Looks like https://github.com/dotnet/runtime/blob/f85c316051f7ba53e54349620e09e034dbd1a812/src/libraries/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs#L54-L57 need to take Android into account to?!

wfurt commented:
something is not right @gfoidl 
```
  [100%] Linking CXX executable singlefilehost
  Stripping symbols from $<TARGET_FILE:hostpolicy> into file $<TARGET_FILE:hostpolicy>.dbg
  [100%] Built target hostpolicy
  /__w/1/s/artifacts/bin/native/net6.0-Linux-Debug-x64/libSystem.Native.a(pal_networking.c.o): In function `SystemNative_GetHostEntryForNameAsync':
  /__w/1/s/src/libraries/Native/Unix/System.Native/pal_networking.c:646: undefined reference to `getaddrinfo_a'
  /__w/1/s/artifacts/bin/native/net6.0-Linux-Debug-x64/libSystem.Native.a(pal_networking.c.o): In function `GetHostEntryForNameAsyncComplete':
  /__w/1/s/src/libraries/Native/Unix/System.Native/pal_networking.c:527: undefined reference to `gai_error'
  clang-9: error: linker command failed with exit code 1 (use -v to see invocation)
  gmake[2]: *** [cli/apphost/static/singlefilehost] Error 1
  gmake[1]: *** [cli/apphost/static/CMakeFiles/singlefilehost.dir/all] Error 2
  gmake: *** [all] Error 2
/__w/1/s/src/installer/corehost/corehost.proj(49,5): error MSB3073: The command "/__w/1/s/src/installer/corehost/build.sh Debug x64 -apphostver "6.0.0-ci" -hostver "6.0.0-ci" -fxrver "6.0.0-ci" -policyver "6.0.0-ci" -commithash "821d4d3b0118f771e101fce7f4bf7c99924773b5" -os Linux -cmakeargs "-DVERSION_FILE_PATH=/__w/1/s/artifacts/obj/_version.c" -coreclrartifacts /__w/1/s/artifacts/bin/coreclr/Linux.x64.Release/ -nativelibsartifacts /__w/1/s/artifacts/bin/native/net6.0-Linux-Debug-x64 -runtimeflavor CoreCLR" exited with code 2.
##[error]src/installer/corehost/corehost.proj(49,5): error MSB3073: (NETCORE_ENGINEERING_TELEMETRY=Build) The command "/__w/1/s/src/installer/corehost/build.sh Debug x64 -apphostver "6.0.0-ci" -hostver "6.0.0-ci" -fxrver "6.0.0-ci" -policyver "6.0.0-ci" -commithash "821d4d3b0118f771e101fce7f4bf7c99924773b5" -os Linux -cmakeargs "-DVERSION_FILE_PATH=/__w/1/s/artifacts/obj/_version.c" -coreclrartifacts /__w/1/s/artifacts/bin/coreclr/Linux.x64.Release/ -nativelibsartifacts /__w/1/s/artifacts/bin/native/net6.0-Linux-Debug-x64 -runtimeflavor CoreCLR" exited with code 2.

```

I'm wondering if this is related to the single-file business.
cc: @VSadov @vitek-karas

VSadov commented on src/libraries/Native/Unix/System.Native/CMakeLists.txt:
```diff
@@ -70,6 +70,10 @@ if (GEN_SHARED_LIB)
     install_with_stripped_symbols (System.Native PROGRAMS .)
+
+    if (HAVE_GETADDRINFO_A)
+        target_link_libraries(System.Native anl)
```
This should be done in extra_libs.cmake, otherwise noone else will know that there is an extra lib involved

VSadov replied:
Also, since it is a CMAKE_REQUIRED_LIBRARIES for Linux, it may be more reliable to check if (CLR_CMAKE_TARGET_LINUX). 

Alternatively, everything that needs to link this will also need to know about `HAVE_GETADDRINFO_A`. 
Thus if this is required on Linux and Linux only, checking for Linux would be simpler.

---
In reply to: [545597019](https://github.com/dotnet/runtime/pull/34633#discussion_r545597019) [](ancestors = 545597019)

VSadov commented on src/libraries/Native/Unix/System.Native/pal_networking.h:
```diff
@@ -301,8 +301,16 @@ typedef struct
 
+PALEXPORT int32_t SystemNative_PlatformSupportsGetAddrInfoAsync(void);
+
 PALEXPORT int32_t SystemNative_GetHostEntryForName(const uint8_t* address, int32_t addressFamily, HostEntry* entry);
```
New public exports will need to be added to the table in  entrypoints.c , so that they can be called in single-file configuration as well - when this is not a separate .so/.dylib.

Just need to add method names to the dispatch table.

There is a build time check that will get upset if the dispatch table and exports get out of sync.

gfoidl replied:
Thanks for the hints!

Just saw your https://github.com/dotnet/runtime/pull/45995 -- is this in the same vein?

gfoidl committed:
Use OperatinngSystem instead of RuntimeInformation in tests

gfoidl committed:
Merge branch 'master' into unix-async-name-resolution

gfoidl committed:
PR Feedback for CMake

gfoidl commented on src/libraries/Native/Unix/System.Native/extra_libs.cmake:
```diff
@@ -13,4 +13,8 @@ macro(append_extra_system_libs NativeLibsExtra)
         list(APPEND ${NativeLibsExtra} "-framework Foundation")
     endif ()
+
+    if (CLR_CMAKE_TARGET_LINUX AND HAVE_GETADDRINFO_A)
```
@VSadov in https://github.com/dotnet/runtime/pull/34633#discussion_r545599929 did you like this or just
```suggestion
    if (CLR_CMAKE_TARGET_LINUX)
```
?

If `CLR_CMAKE_TARGET_LINUX` is true, then `HAVE_GETADDRINFO_A` will be true too for libraries-build here?
When that's the case https://github.com/dotnet/runtime/blob/33da8da9dd3c2ceced296876af8bd9189cc3741f/src/libraries/Native/Unix/configure.cmake#L904-L915 could be simplified too.

Now I have it this way, to be on the safe side.

VSadov replied:
Single file host builds in installer partition. It will be moved to clr eventually, but either way it is not in libraries. 
So either `HAVE_GETADDRINFO_A` probe needs to be done in other partitions too, or we could do just `if (CLR_CMAKE_TARGET_LINUX)`

I think `if (CLR_CMAKE_TARGET_LINUX)` would be sufficient. 
Since you are making `anl` a `CMAKE_REQUIRED_LIBRARIES`, we would not even get this far if it is not present.

gfoidl replied:
> Single file host builds in installer partition.

That's new information for me -- thanks!

gfoidl commented:
[CI fails](https://dev.azure.com/dnceng/public/_build/results?buildId=926308&view=logs&jobId=000ba8e3-34f8-51d2-8d06-9b61f10256d0&j=000ba8e3-34f8-51d2-8d06-9b61f10256d0&t=ca9b6e26-8c16-57f9-673e-2bc0dc655f25) now with
```
...
[ 99%] Building CXX object cli/hostpolicy/standalone/CMakeFiles/hostpolicy.dir/__/__/fxr/fx_ver.cpp.o
  [ 99%] Building CXX object cli/hostpolicy/standalone/CMakeFiles/hostpolicy.dir/__/__/hostmisc/pal.unix.cpp.o
  [100%] Linking CXX shared library libhostpolicy.so
  Stripping symbols from $<TARGET_FILE:hostpolicy> into file $<TARGET_FILE:hostpolicy>.dbg
  [100%] Built target hostpolicy
  /__w/1/s/artifacts/bin/coreclr/Linux.x64.Debug//lib/libSystem.Native.a(pal_networking.c.o): In function `SystemNative_GetHostEntryForNameAsync':
  /__w/1/s/src/libraries/Native/Unix/System.Native/pal_networking.c:646: undefined reference to `getaddrinfo_a'
  /__w/1/s/artifacts/bin/coreclr/Linux.x64.Debug//lib/libSystem.Native.a(pal_networking.c.o): In function `GetHostEntryForNameAsyncComplete':
  /__w/1/s/src/libraries/Native/Unix/System.Native/pal_networking.c:527: undefined reference to `gai_error'
  clang-9: error: linker command failed with exit code 1 (use -v to see invocation)
  gmake[2]: *** [cli/apphost/static/singlefilehost] Error 1
  gmake[1]: *** [cli/apphost/static/CMakeFiles/singlefilehost.dir/all] Error 2
  gmake: *** [all] Error 2
/__w/1/s/src/installer/corehost/corehost.proj(49,5): error MSB3073: The command "/__w/1/s/src/installer/corehost/build.sh Debug x64 -apphostver "6.0.0-ci" -hostver "6.0.0-ci" -fxrver "6.0.0-ci" -policyver "6.0.0-ci" -commithash "af4ae1ef7450fba071f352539009a335c99933ad" -os Linux -cmakeargs "-DVERSION_FILE_PATH=/__w/1/s/artifacts/obj/_version.c" -coreclrartifacts /__w/1/s/artifacts/bin/coreclr/Linux.x64.Debug/ -nativelibsartifacts /__w/1/s/artifacts/bin/native/net6.0-Linux-Debug-x64 -runtimeflavor CoreCLR" exited with code 2.
##[error]src/installer/corehost/corehost.proj(49,5): error MSB3073: (NETCORE_ENGINEERING_TELEMETRY=Build) The command "/__w/1/s/src/installer/corehost/build.sh Debug x64 -apphostver "6.0.0-ci" -hostver "6.0.0-ci" -fxrver "6.0.0-ci" -policyver "6.0.0-ci" -commithash "af4ae1ef7450fba071f352539009a335c99933ad" -os Linux -cmakeargs "-DVERSION_FILE_PATH=/__w/1/s/artifacts/obj/_version.c" -coreclrartifacts /__w/1/s/artifacts/bin/coreclr/Linux.x64.Debug/ -nativelibsartifacts /__w/1/s/artifacts/bin/native/net6.0-Linux-Debug-x64 -runtimeflavor CoreCLR" exited with code 2.

Build FAILED.

/__w/1/s/src/installer/corehost/corehost.proj(49,5): error MSB3073: The command "/__w/1/s/src/installer/corehost/build.sh Debug x64 -apphostver "6.0.0-ci" -hostver "6.0.0-ci" -fxrver "6.0.0-ci" -policyver "6.0.0-ci" -commithash "af4ae1ef7450fba071f352539009a335c99933ad" -os Linux -cmakeargs "-DVERSION_FILE_PATH=/__w/1/s/artifacts/obj/_version.c" -coreclrartifacts /__w/1/s/artifacts/bin/coreclr/Linux.x64.Debug/ -nativelibsartifacts /__w/1/s/artifacts/bin/native/net6.0-Linux-Debug-x64 -runtimeflavor CoreCLR" exited with code 2.
    0 Warning(s)
    1 Error(s)
```

Please help me -- a local build (in a linux-vm `./build.sh clr+libs -rc Release`, although not a clean build) ran fine.

VSadov committed:
Update src/libraries/Native/Unix/System.Native/extra_libs.cmake

Co-authored-by: Günther Foidl <gue@korporal.at>

VSadov commented:
It looks like `anl` is not necessarily present on Linux. I see the failures at least on Linux_musl and Android where the build could not find the library. That is a bit inconvenient.

One way to deal with this is to make the library optional and make the feature conditional on `HAVE_GETADDRINFO_A`.
That would imply making `HAVE_GETADDRINFO_A` work in installer partition.

wfurt commented:
yes, this is part of glibc - so it will not be there in MUSL based distributions @VSadov . I look as I was concerned about extra dependency but I think it is ok to use swhen part of glibc bundle.

VSadov commented:
If you go the route of making this conditional on Linux, you would need to add similar probe for `HAVE_GETADDRINFO_A` to `installer/corehost/cli/apphost/static`.

We should think about reducing duplication like this going forward, but for the time being it could be acceptable solution.

wfurt commented:
Do you have plan how to move forward @gfoidl ? (I'm not pushing but making sure you are not stuck) It seems like this got more complicated beyond base DNS changes ;(

gfoidl commented:
### State of the PR

* code changes to native and managed side are done
* tests added
* a follow-up PR is ready to address cancellation
* local build and test on ubuntu succeedes

### Open questions (from review)

* https://github.com/dotnet/runtime/pull/34633/files#r544521828
* https://github.com/dotnet/runtime/pull/34633/files#r406444992

### What's left?

Async name resolution on Unix relies on _glibc_-functionality (`getaddrinfo_a` from lib `anl`). It won't work on distros that don't provide these symbols, hence we check for this and emit the symbol `HAVE_GETADDRINFO_A`.

Here comes the trouble with "single file host builds in installer partition" and for me it's hard to follow what's going on, since this is pretty much down the entrails of .NET 😉 

From https://github.com/dotnet/runtime/pull/34633#discussion_r546006416:
What is meant by "other partitions"?

If I read https://github.com/dotnet/runtime/pull/34633#issuecomment-748268040 correct, so the check for `HAVE_GETADDRINFO_A` should be done in [configure.cmake](https://github.com/dotnet/runtime/blob/master/src/installer/corehost/cli/apphost/static/configure.cmake) too and [CMakeLists.txt](https://github.com/dotnet/runtime/blob/master/src/installer/corehost/cli/apphost/static/CMakeLists.txt) updated to conditionally link based on `HAVE_GETADDRINFO_A`?
Also undo https://github.com/dotnet/runtime/pull/34633/commits/fa9c6f90b7ef5033e3c30a46dfe34dc47d5ef91e

gfoidl committed:
Revert "Update src/libraries/Native/Unix/System.Native/extra_libs.cmake"

This reverts commit fa9c6f90b7ef5033e3c30a46dfe34dc47d5ef91e.

gfoidl committed:
Another to build single file host

Based on https://github.com/dotnet/runtime/pull/34633#issuecomment-756045320

gfoidl commented:
@wfurt good news: CI is green 🎉  -- PTAL

wfurt reviewed (approved):
LGTM @gfoidl. thanks for the contribution. 

I caused one more merge conflict by rushing fix to servicing. I think that should be simple to resolve. 
I will merge as soon as the PR is green again (and sorry for the troubles)

VSadov reviewed (approved):
LGTM for the cmake changes

gfoidl committed:
Merge branch 'master' into unix-async-name-resolution

gfoidl commented:
[iOS.Simulator.Aot.Test Work Item](https://dev.azure.com/dnceng/public/_build/results?buildId=950139&view=ms.vss-test-web.build-test-results-tab) fails due to this test [TryGetAddrInfo_EmptyHost](https://github.com/gfoidl/dotnet-runtime/blob/89a7c5a03b35bfbdb52123cf1848bc5039e947fb/src/libraries/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs#L51-L59).

The check for the operating system got taken over from existing tests, which have the comment
```c#
// On Unix, we are not guaranteed to be able to resove the local host. The ability to do so depends on the
// machine configurations, which varies by distro and is often inconsistent.
```
As more and more different OS come into play, maybe it's safer to flip this check and test for `!Windows` here?
Is there a better way to handle this fact, except just returning from the test-code?

gfoidl committed:
Test for !Windows instead excluding several Unix-flavors

Cf. https://github.com/dotnet/runtime/pull/34633#issuecomment-760205332

gfoidl commented:
@wfurt CI looks green 😃 
Please have a look at the last commit (https://github.com/dotnet/runtime/pull/34633/commits/dd1e2453c70d20d2be664c7f0041be7b724113e3) -- I'm not entirely happy with this.

For the test on windows I'll follow https://github.com/dotnet/runtime/pull/34633#discussion_r557004419 and create a separate PR for this, so it can be handled in isolation.
Also there will come a PR to add cancellation-support.

karelz commented:
@gfoidl thanks a lot for your PR!!! Also, appreciate your patience with this one ;)

gfoidl commented:
This was an interesting journey with some very valuable input (some C hints, and being down at the PAL to see how things work / a place where a usual dotnet-developer won't be), so I'd say: perfect (except that it took that long, but now it's in so this doesn't matter).
